### DISCUSSION TITLE
New experimental `git_status_async` option

### DISCUSSION THREAD

**Discussion by cseickel** [nvim-neo-tree/neo-tree.nvim#152] [Announcements]
I just added in the ability to run the git status check as a background check instead of as part of every render. I'm adding this in as an undocumented option right now because I want to test it out before making it the new default. If it works as expected, I'll just drop the undocumented option and always run this process async.

Along with that is another new option to run the gitignore checks perf rolder as needed instead of for the whole repo in one shot. In that case, which one is more efficient probably depends on the size and speed of your repo.

I'm looking for testers! Here are the options to try out the new functionality:

```lua
      require("neo-tree").setup({
        enable_git_status = true,
        git_status_async = true, -- <-- this is the new setting
        filesystem = {
          filters = {
            respect_gitignore = true,
            gitignore_source = "git check-ignore", -- <-- this is the other new setting, it runs a per folder check for each folder that is open.
            --gitignore_source = "git status", -- <-- this is the current default it reads the ignored files for the whole repo in one command
          }
        }
      })
```

if you try it out, please provide feedback. Thanks!

**Comment by nyngwang** [comment]
Excited about this upcoming optimization :) Currently, if I open the tree via the toggle on a large repo., there will be a sensible delay. (not sure whether this is related to this upcoming feature though. been too busy these days so I might have missed many important announcements.)

> [...] probably depends on the size and speed of your repo. (p1)

bwt, What did you mean speed of a repo.?

**Reply by cseickel** [reply]
> Excited about this upcoming optimization :) 

It's not an _upcoming_ optimization, it's here now! You just need to opt into it because I still consider it experimental.



> > [...] probably depends on the size and speed of your repo. (p1)
> 
> bwt, What did you mean speed of a repo.?

I meant how fast git operations run within your repo. Sometimes something like a `git status` runs in milliseconds, sometimes it takes several seconds. This generally depends on how many files are being tracked, but it is also affected by cache problems, anti-virus, slow hard drives, etc.

**Reply by nyngwang** [reply]
Got it :) Let me try it now!

**Comment by cseickel** [comment]
I probably should have mentioned that this was in **main** when I wrote the announcement, but has just been released to **v1.x** / **v1.38** now.
